ipynb/android-wloads/Android_YoutTube_Tests.ipynb

```diff
@@ -0,0 +1,500 @@
+    "    target.cpufreq.set_all_governors('ondemand')\n",
+    "    \n",
+    "    for cpu in target.list_online_cpus():\n",
+    "        tunables = target.cpufreq.get_governor_tunables(cpu)\n",
+    "        target.cpufreq.set_governor_tunables(\n",
+    "            cpu,\n",
+    "            'ondemand',\n",
+    "            **{'sampling_rate' : tunables['sampling_rate_min']}\n",
+    "        )"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "# CPUFreq configurations to test\n",
+    "confs = {\n",
+    "    'performance' : {\n",
+    "        'label' : 'prf',\n",
+    "        'set' :  set_performance,\n",
+    "    },\n",
+    "    'powersave' : {\n",
+    "        'label' : 'pws',\n",
+    "        'set' :  set_powersave,\n",
+    "    },\n",
+    "     'interactive' : {\n",
+    "         'label' : 'int',\n",
+    "         'set' :  set_interactive,\n",
+    "     },\n",
+    "    'sched' : {\n",
+    "        'label' : 'sch',\n",
+    "        'set' :  set_sched,\n",
+    "    },\n",
+    "    'ondemand' : {\n",
+    "        'label' : 'odm',\n",
+    "        'set' :  set_ondemand,\n",
+    "    }\n",
+    "}\n",
+    "\n",
+    "# The set of results for each comparison test\n",
+    "results = {}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "def youtube_run(exp_dir, video_url, video_duration_s):\n",
+    "    # Unlock device screen (assume no password required)\n",
+    "    target.execute('input keyevent 82')\n",
+    "    # Press Back button to be sure we run the video from the start\n",
+    "    target.execute('input keyevent KEYCODE_BACK')\n",
+    "\n",
+    "    # Start YouTube video on the target device\n",
+    "    target.execute('am start -a android.intent.action.VIEW \"{}\"'.format(video_url))\n",
+    "    # Allow the activity to start\n",
+    "    sleep(3)\n",
+    "    # Reset framestats collection\n",
+    "    target.execute('dumpsys gfxinfo com.google.android.youtube reset')\n",
+    "    # Wait until the end of the video\n",
+    "    sleep(video_duration_s)\n",
+    "    \n",
+    "    # Get frame stats\n",
+    "    framestats_file = os.path.join(exp_dir, \"framestats.txt\")\n",
+    "    adb_command(target.adb_name, 'shell dumpsys gfxinfo com.google.android.youtube > {}'.format(framestats_file))\n",
+    "\n",
+    "    # Close application\n",
+    "    target.execute('am force-stop com.google.android.youtube')\n",
+    "\n",
+    "    # Clear application data\n",
+    "    target.execute('pm clear com.google.android.youtube')\n",
+    "\n",
+    "    return framestats_file"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "def experiment(governor, exp_dir):\n",
```

Comment:
To run using SysTrace we need a couple of define in a cell before the definition of execute.
I usually have these define at the beginning of the document, right before the target/test configuration

```python
CATAPULT_HOME = '~/Code/catapult'
SYSTRACE_CMD  = CATAPULT_HOME + "/systrace/systrace/systrace.py -o {} gfx view sched -t {}"
```

The experiment function can be re-defined to use either FTrace or SysTrace that way:

```python
def experiment(governor, tcmd, exp_dir, collect='all'):
    results = {}
    
    logging.info("Experiment: governor=%s, collect=%s", governor, collect)
    
    # Ensure all CPUs are online
    #target.execute('echo 4 2 > /proc/hps/num_base_perf_serv')
    
    os.system('mkdir -p {}'.format(exp_dir));
    results['exp_dir'] = exp_dir
    
    logging.info('------------------------')
    logging.info('Run workload: %s', tcmd)

    # Setup CPUFreq governor
    confs[governor]['set']()
    logging.info('    governor:\n%s',
                 json.dumps(target.cpufreq.get_all_governors(), indent=4))

    # Start the required tracing command
    if 'all' in collect or 'ftrace' in collect:
        # Start FTrace and Energy monitoring
        te.ftrace.start()
    elif 'systrace' in collect:
        # Start systrace
        results['trace_file'] = os.path.join(exp_dir, 'trace.html')
        systrace_output = pe.spawn(SYSTRACE_CMD.format(
                                    results['trace_file'], SYSTRACE_TIME))

    # Execte the workload
    output = target.execute(tcmd)

    # Stop the required trace command
    if 'all' in collect or 'ftrace' in collect:
        te.ftrace.stop()
    elif 'systrace' in collect:
        logging.info('Waiting systrace report [%s]...', results['trace_file'])
        systrace_output.wait()
    
    if 'all' in collect or 'stats' in collect:
        # Processing test resuts
        logging.info('Parsing experiment results...')
        df = dataframe(output)
        results['df'] = df
    
        # Computing stats
        logging.info('Computing execution stats...')
        results['stats'] = stats(df)
    
    # Dump platform description
    te.platform_dump(exp_dir)
    
    # Collect and keep track of the trace
    if 'all' in collect or 'ftrace' in collect:
        logging.info('Collecting FTrace events...')
        trace_file = os.path.join(exp_dir, 'trace.dat')
        te.ftrace.get_trace(trace_file)
        results['trace_file'] = trace_file        
    
    # Parse trace
    if 'all' in collect or 'trace_analysis' in collect:
        logging.info('Trace analysis...')
        tr = Trace(te.platform, results['trace_file'],
                   events=my_conf['ftrace']['events'],
                   trace_format='SysTrace')
        results['trace'] = tr
        results['ftrace'] = tr.ftrace
        results['ta'] = TraceAnalysis(tr, te.platform)
    
        # return all the experiment data
    return results
```

where basically we have a "collect" attribute which can be used to defined if we want to use FTrace or SysTrace.
The tracing time is also exposed as a parameter, which is used to define the SysTrace length as well as the video duration.

**NOTE** I've changed the video to be the popular "Big Buck Bunny" video "benchmark"... it has the benefit of playing always without advertisements in front... donno if it's intentional... maybe I was just luky ;-)